Create new methods to parse variables based on their IDs

b1quint opened the pull request:
Zephyr Scale and Jira contain several variables in their database that are returned as JSON objects inside the GET requests' response. This makes it harder to tie the payload with the frontend quickly. This PR ensures that we have all the relevant variables parsed correctly in the API side. 

As part of the implementation of these methods to parse information, I will create a command line interface that we can use to easily confirm how the payload is set.

b1quint committed:
In ``zephyr_scale.py``, implement new method ``get(...)``

The code had lots of duplication.
This new method is the resposible for the actuall communication
from/to Zephyr.

# Conflicts:
#	python/lsst/ts/planning/tool/zephyr_interface.py

b1quint committed:
Start new command line interface for the Zephyr Scale API

This new command line interface (cli) uses argparse to parse
commands from the command line. It is now used to get test cases
only. But it is design to incorporate more functionality.

As part of the development, I updated the ``get_test_case`` method
to receive the ``raw`` argument. If it is false, some fields in the
payload will be parsed into more human readable format to easily
display in the love player tool

b1quint committed:
Add new option to CLI to get test steps from a test case

In addition, add new option to query user based on its hash code.

b1quint committed:
Add new option to CLI to get a test cycle

In addition, the ``get_test_cycle`` method can return a raw json
string if the new argument ``raw`` is true.

b1quint committed:
Add new CLI to extract a test execution and add new method parse

b1quint committed:
Big refactor in favour of the ``parse`` method

We had five methods with lots of duplicated code.
I replaced all of them with the ``parse`` method.
Further stripping is now removed from the function.
This allows an easier access to the api to be explored if we want to.

b1quint committed:
Update get_test_cycle and get_test_case

In addition, update test cases and CLI interface to match.
Add new json files with examples of simplified payloads.

b1quint committed:
Update get_test_execution to parse outputs

sebastian-aranda commented on README.md:
```diff
@@ -1,32 +1,73 @@
 It has been tested with Jira Cloud and Zephyr Scale Cloud.
-You will need an API token from both for it to work. 
+You will need API tokens from both for it to work.
+Here are the list of environment variables required:
```
```suggestion
Here is the list of environment variables required:
```

b1quint committed:
Rename get_test_executions to list_test_executions

For two reasons: first, to avoid mixing up get_test_execution with
get_test_executions. Second, to have a second category of queries
where we want a list of things. In addition, update zapi with new
command `list` and `test_executions`.

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
```diff
@@ -93,31 +94,25 @@ def __init__(
         """
-        Extracts the test case name and version from the given test execution
-        JSON.
+        Generic method to parse get requests to the Zephyr Scale API.
```
This is not really parsing, so perhaps this should be something like:
```suggestion
        Generic method to make HTTP GET requests to the Zephyr Scale API and return the payload.
```

sebastian-aranda commented on python/lsst/ts/planning/tool/cli.py:
```diff
@@ -0,0 +1,192 @@
+    jira_username = os.getenv("JIRA_USERNAME")
+
+    if zephyr_token is None:
+        raise ValueError("ZEPHYR_TOKEN environment variable not set.")
```
```suggestion
        raise ValueError("ZEPHYR_API_TOKEN environment variable not set.")
```

sebastian-aranda commented on python/lsst/ts/planning/tool/cli.py:
```diff
@@ -0,0 +1,192 @@
+    if zephyr_token is None:
+        raise ValueError("ZEPHYR_TOKEN environment variable not set.")
+    if jira_token is None:
+        raise ValueError("JIRA_TOKEN environment variable not set.")
```
```suggestion
        raise ValueError("JIRA_API_TOKEN environment variable not set.")
```

sebastian-aranda commented on python/lsst/ts/planning/tool/cli.py:
```diff
@@ -0,0 +1,192 @@
+    parse_test_cycle.add_argument("test_cycle_key", type=str)
+    parse_test_cycle.add_argument("--indent", type=int, default=4)
+    parse_test_cycle.add_argument(
+        "--parse", choices=["raw", "full", "simple"], default="raw"
```
You missed the `-p` option here:
```suggestion
        "-p", "--parse", choices=["raw", "full", "simple"], default="raw"
```

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
```diff
@@ -233,25 +245,34 @@ async def get_test_cycle(self, test_cycle_key):
+
+        return test_cycle
 
     async def get_test_cycles(
```
Perhaps you could add a CLI method to query test cycles such as it happens with the `list_test_executions` method.

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
```diff
@@ -164,6 +167,26 @@ async def test_get_test_cycle(self):
+            "isLast",
+            "values",
+        ]
+        payload_value_keys = ["inline", "testCase"]
```
You could also check the payload of the `inline` object, as the main purpose of the method is to bring the data of the test steps (which basically come inside the `inline` object).

sebastian-aranda commented on tests/test_zephyr_interface_no_ci.py:
The following test: `test_get_test_executions` is calling `zapi.get_test_executions`, but that method doesn't exist anymore.

sebastian-aranda replied:
On a separate note: we should try hooking unit tests to the CI, which doesn't seem to be running. We will probably need to hook up the library with `ts_jenkins_shared_library`. This can be probably done on a separate PR.

sebastian-aranda commented on tests/test_zephyr_interface.py:
The following test: `test_get_test_executions` is calling `zapi.get_test_executions`, but that method doesn't exist anymore.

b1quint replied:
This should be solved now

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
```diff
@@ -157,70 +144,95 @@ async def get_test_case(self, test_case_key):
+        if parse == "raw":
+            return test_case
+
+        parse_fields = {
+            "project": "key",
+            "priority": "name",
+            "status": "name",
         }
```
I would also add a couple other fields to parse: `component` and `folder`.

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
```diff
@@ -411,55 +459,86 @@ async def get_test_executions(self, test_cycle_key, max_results=20):
         """
-        Get the user name based on a JSON object containing "self"
-        and "accountId" keys.
+        Get the a user name based on a json_obj payload containing `self` and
```
```suggestion
        Get the user name based on a json_obj payload containing `self` and
```

sebastian-aranda commented on python/lsst/ts/planning/tool/zephyr_interface.py:
```diff
@@ -411,55 +459,86 @@ async def get_test_executions(self, test_cycle_key, max_results=20):
-            The account ID of the user.
+        user : dict or str
+            This can be a dictionary containing the `self` and `accountId` keys
+            representing a user or  a single string containng the `accountId`.
```
```suggestion
            representing a user or a single string containng the `accountId`.
```

sebastian-aranda reviewed (changes requested):
Nice work @b1quint. I took some time to try integrating the full refactor you did. I like how the code is designed. I've left some comments and requests, please take a look when convenient. After that I'll try jumping into adding a couple methods that I don't see now:
- Query all test cases for a given test cycle + their last executions. (to display the sidebar panel with the list of test cases).
- Query/Gather all test executions for a given test cycle **AND** test case.

b1quint committed:
Apply suggestions from PR#4

Those are minor typo suggestions and errors in the code.

b1quint committed:
Update unit tests to target ``list_test_executions`` method

sebastian-aranda reviewed (approved):
Thanks for all the efforts put in here @b1quint ! PR approved :+1:.
